### PR TITLE
Eliminate the need for inheritance for action controllers.

### PR DESCRIPTION
**Pull request by Vinai**
### Description
Kindly see issue https://github.com/magento/community-features/issues/9 for detailed information.

### Fixed Issues (if relevant)
All the CE unit and integration tests pass. The big question is how the EE will do.
When this PR is merged, frontend action controllers don't have to extend AbstractAction, they only have to implement the ActionInterface.

This PR mainly moves the dispatching of the controller_predispatch events into a plugin.  
It also changes the plugins for `Action::dispatch()` method to `ActionInterface::execute()`.  
Finally, it preserves the behavior in regards to the `NO_DISPATCH` and `NO_POST_DISPATCH` action flags.

Backend action controllers still have to extend the backend module abstract Action for authentication - moving that to a plugin will be a separate PR.

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by Vinai** [comment]
The [failing Codacity quality review](https://www.codacy.com/app/vrann/magento2/pullRequest?prid=1213377) is a false positive, since using the object manager in integration tests is necessary.
Is there an annotation that can be used to fix that?
EDIT: I'm trying to use `@SuppressWarnings(PHPMD.StaticAccess)`.

**Comment by dmanners** [comment]
@Vinai thanks for this and for fixing the false positive. I have message @antonkril about taking a look over this PR since he was the first one that raised this problem.

**Comment by Vinai** [comment]
Did supressing the check work? Anyway, if there’s something I can do to make the PR better so it can be accepted let me know please.

**Comment by Vinai** [comment]
Adding the `@SuppressWarnings(PHPMD.StaticAccess)` annotation to the method PHPDoc block did not suppress the warning. I've amended the commit with the annotation on the whole class - maybe that works in Codacity.

File: app/code/Magento/Customer/Test/Unit/Model/App/Action/ContextPluginTest.php
Comment (paliarush) [inline app/code/Magento/Customer/Test/Unit/Model/App/Action/ContextPluginTest.php]: `beforeExecute` does not accept second argument.

File: app/code/Magento/Store/Test/Unit/App/Action/Plugin/StoreCheckTest.php
Comment (paliarush) [inline app/code/Magento/Store/Test/Unit/App/Action/Plugin/StoreCheckTest.php]: `beforeExecute` does not accept second argument

File: app/code/Magento/Theme/Test/Unit/Model/Theme/Plugin/RegistrationTest.php
Comment (paliarush) [inline app/code/Magento/Theme/Test/Unit/Model/Theme/Plugin/RegistrationTest.php]: Request argument should be removed.

File: app/code/Magento/Weee/Test/Unit/App/Action/ContextPluginTest.php
Comment (paliarush) [inline app/code/Magento/Weee/Test/Unit/App/Action/ContextPluginTest.php]: Second argument should be removed (including other occurrences below)

File: lib/internal/Magento/Framework/App/Action/Plugin/ActionFlagNoDispatchPlugin.php
Comment (paliarush) [inline lib/internal/Magento/Framework/App/Action/Plugin/ActionFlagNoDispatchPlugin.php]: Please add php doc

File: lib/internal/Magento/Framework/App/Action/Plugin/EventDispatchPlugin.php
Comment (paliarush) [inline lib/internal/Magento/Framework/App/Action/Plugin/EventDispatchPlugin.php]: Please add php doc

File: lib/internal/Magento/Framework/App/Action/Plugin/EventDispatchPlugin.php
Comment (paliarush) [inline lib/internal/Magento/Framework/App/Action/Plugin/EventDispatchPlugin.php]: Should return void when arguments are not modified.

**Review by paliarush — CHANGES_REQUESTED** [review]
1. Please add php docs for all new methods and classes (except for tests)

2. The following plugins should be rewritten as well:
CE: `\Magento\Customer\Model\Plugin\CustomerNotification::beforeDispatch`
EE: `\Magento\Logging\App\Action\Plugin\Log::beforeDispatch`
`\Magento\CustomerSegment\Model\App\Action\ContextPlugin::aroundDispatch`
B2B: `\Magento\Company\Plugin\Framework\App\Action\AbstractActionPlugin::aroundDispatch`
CE should be fixed in scope of current PR. 
Related PRs should be created for EE and B2B.

3. Migration steps should be added to the list of backward incompatible changes on the dev docs for someone who declares plugins for `dispatch()`

4. What do you think about deprecating `\Magento\Framework\App\Action\AbstractAction::dispatch` and not invoking it anymore from `\Magento\Framework\App\FrontController::dispatch` (after moving authentication)?

5. In the PhpDoc of `\Magento\Framework\App\ActionInterface::execute()` we promise to add 'request' as an argument in the future. I don't think it will be possible because of backward compatibility issues for all existing action controllers. Do you see any solution for this? If not, let's remove that comment

File: app/code/Magento/Customer/Test/Unit/Model/App/Action/ContextPluginTest.php
Comment (Vinai) [reply app/code/Magento/Customer/Test/Unit/Model/App/Action/ContextPluginTest.php]: Fixed

File: app/code/Magento/Store/Test/Unit/App/Action/Plugin/StoreCheckTest.php
Comment (Vinai) [reply app/code/Magento/Store/Test/Unit/App/Action/Plugin/StoreCheckTest.php]: Fixed

File: app/code/Magento/Theme/Test/Unit/Model/Theme/Plugin/RegistrationTest.php
Comment (Vinai) [reply app/code/Magento/Theme/Test/Unit/Model/Theme/Plugin/RegistrationTest.php]: Fixed.

File: app/code/Magento/Weee/Test/Unit/App/Action/ContextPluginTest.php
Comment (Vinai) [reply app/code/Magento/Weee/Test/Unit/App/Action/ContextPluginTest.php]: Fixed.

File: lib/internal/Magento/Framework/App/Action/Plugin/ActionFlagNoDispatchPlugin.php
Comment (Vinai) [reply lib/internal/Magento/Framework/App/Action/Plugin/ActionFlagNoDispatchPlugin.php]: What goes there? Looking at other classes I find not all have a PHPDoc class comment. It only seems to be present if there is an `@api` annotation or a `@SuppressWarnings` annotation.
Can you point me to some documentation maybe?

File: lib/internal/Magento/Framework/App/Action/Plugin/EventDispatchPlugin.php
Comment (Vinai) [reply lib/internal/Magento/Framework/App/Action/Plugin/EventDispatchPlugin.php]: I have the same question as for the `ActionFlagNoDispatchPlugin` above.

File: lib/internal/Magento/Framework/App/Action/Plugin/EventDispatchPlugin.php
Comment (Vinai) [reply lib/internal/Magento/Framework/App/Action/Plugin/EventDispatchPlugin.php]: Fixed.

**Comment by Vinai** [comment]
Thank you for your review, @paliarush!  

Regarding your non-inline comments I have a couple of questions:

> 1. Please add php docs for all new methods and classes (except for tests)

I think I've added PHPDoc blocks to all non-test methods. What do I set for classes that don't have `@api` or `@SuppressWarnings` annotations? I can't find consistent examples using the core code.

> 3. Migration steps should be added to the list of backward incompatible changes on the dev docs for someone who declares plugins for `dispatch()`

I don't think there are any backward incompatible changes introduced by this PR.

> 4. What do you think about deprecating `\Magento\Framework\App\Action\AbstractAction::dispatch` and not invoking it anymore from `\Magento\Framework\App\FrontController::dispatch` (after moving authentication)?

That would be a nice change. Not calling `dispatch` any more would break soooooo much third party code, I think it will probably have to stay in place for a very long time.
That said, I would love adding a `@deprecated` annotation to `\Magento\Framework\App\Action\AbstractAction::dispatch`. But is this PR the right scope for that? I think probably not.

> 5. In the PhpDoc of `\Magento\Framework\App\ActionInterface::execute()` we promise to add 'request' as an argument in the future. I don't think it will be possible because of backward compatibility issues for all existing action controllers. Do you see any solution for this? If not, let's remove that comment.

I don't see a way to do that either without causing massive disruption, and that change is not worth it. A consequence of that is that the request object should be declared an injectable (which is how it is already used in much third party code anyway). This is only a devdocs change I guess.

I'm working on patching `\Magento\Customer\Model\Plugin\CustomerNotification::beforeDispatch` and will update the PR when I'm done.

**Comment by Vinai** [comment]
Ready for next review (if the build is green).

**Comment by Vinai** [comment]
`\Magento\Customer\Model\Plugin\CustomerNotification::beforeDispatch()` refactored and the static tests are fixed. Ready for next review.

**Comment by paliarush** [comment]
@Vinai Thank you for such a valuable PR.

1. I meant adding some description to PhpDoc for every new method and class, not just annotations.

2. So how do we make sure that corresponding PRs to EE and B2B will be merged along with current one? This requirement should be made clear for Community Engineering team, before they merge

3. This PR is backward incompatible from SPI point of view. Scenario: someone creates a new plugin for `dispatch()` in their extension. After this PR, such customizations will be ignored for all controllers not extending from the abstract controller. That's why we must make sure this will be highlighted in the  backward compatibility guide

4. Could you please explain the scenario when it will break 3rd party extensions (assuming we accept corresponding PR for the backend controllers)? I believe it is a good PR to mark `dispatch()` as deprecated

5. Please remove the promise from DocBlock of `\Magento\Framework\App\ActionInterface::execute()`

**Comment by Vinai** [comment]
> 1. I meant adding some description to PhpDoc for every new method and class, not just annotations.

Thanks for clarifying, done.

> 2. how do we make sure that corresponding PRs to EE and B2B will be merged along with current one? This requirement should be made clear for Community Engineering team, before they merge.

I don't know. Do you have a suggestion how to proceed? Maybe it can be coordinated via slack?

> 3. This PR is backward incompatible from SPI point of view. Scenario: someone creates a new plugin for dispatch() in their extension. After this PR, such customizations will be ignored for all controllers not extending from the abstract controller. That's why we must make sure this will be highlighted in the backward compatibility guide

This already is the case since 2.0. Any frontend action that doesn't return a `Layout` or `Page` result already can simply implement the `ActionInterface` and it works just fine. I know of several cases where the `Action` class isn't used. In those cases `dispatch()` already isn't called.
But okay, I see your point - probably many developers are not aware of this.
On what devdocs page are such changes collected? What branch? If you want I can open a PR.

> 4. Could you please explain the scenario when it will break 3rd party extensions (assuming we accept corresponding PR for the backend controllers)? I believe it is a good PR to mark dispatch() as deprecated

Marking the `dispatch` method as deprecated will not break 3rd party extensions, just adding the request as an argument to `ActionIngerface::execute()` would.   
When this PR is merged I hope to make another PR that moves the backend authentication into plugins. That would probably be the right PR to mark the `dispatch()` method as `deprecated()`.

> 5. Please remove the promise from DocBlock of `\Magento\Framework\App\ActionInterface::execute()`

Done.

**Comment by Vinai** [comment]
The travis build is hanging. Can you restart it please? I think it is green.

**Comment by paliarush** [comment]
@Vinai thank you for making changes, PR looks good to me now.

@ishakhsuvarov please take care of items 2 and 3 from the list above before merging the PR.

**Comment by mcspronko** [comment]
Hi @Vinai, I've merged the latest mainline with new integration tests, checked and fixed them. Also, fixed PHPDoc for the EventDispatchPlugin class.

**Comment by Vinai** [comment]
Thanks @mcspronko.

@paliarush, I think I addressed all changes you requested, yet the review is still marked as pending? Can you check please?

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=13045) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=13045) before we can accept your contribution.<br/>**7** out of **13** committers have signed the CLA.<br/><br/>:white_check_mark: VladimirZaets<br/>:white_check_mark: VitaliyBoyko<br/>:white_check_mark: chirag-wagento<br/>:white_check_mark: hitesh-wagento<br/>:white_check_mark: NamrataChangani<br/>:white_check_mark: Radio<br/>:white_check_mark: rogyar<br/>:x: wert2all<br/>:x: aakimov<br/>:x: ameysar<br/>:x: magento-engcom-team<br/>:x: Mariana Lashch<br/>:x: olysenko<hr/>**Mariana Lashch** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

**Comment by Vinai** [comment]
I've resolve the merge conflicts. How to proceed with this PR?

**Comment by Vinai** [comment]
Why is the Travis build not starting?

**Comment by Vinai** [comment]
Closing in favor of PR #16268 to the 2.3-develop branch.

**Pull request closed without merging**
